common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+     */
+    ~FileDescriptorImpl() override { close(); }
+
+    /**
+     * @brief Gets file descriptor.
+     *
+     * File descriptor can be used:
+     * - in select(), poll() or epoll() to wait for events on this descriptor.
+     * - in send() family of functions to send data through the socket connection.
+     * - in recv() family of functions to receive data from the socket connection.
+     *
+     * @return File descriptor value.
+     */
+    int fd() override { return m_fd; }
+
+    /**
+     * @brief Closes file descriptor.
+     *
+     * If valid, closes file descriptor and then invalidates it.
+     *
+     * @return True on success and false otherwise (for example, if it was already closed)
+     */
+    bool close() override
+    {
+        printf("close(%d)\n", m_fd);
+        if (FileDescriptor::invalid_descriptor == m_fd) {
+            return false;
+        }
+
+        int rc = ::close(m_fd);
+        m_fd   = FileDescriptor::invalid_descriptor;
+
+        return (0 == rc);
+    }
+
+private:
+    int m_fd;
+};
+
+/**
+ * Wrapper class around sockadr (structure describing a generic socket address)
+ */
+class SockAddr {
+public:
+    virtual ~SockAddr()                             = default;
+    virtual const struct sockaddr *sockaddr() const = 0;
+    virtual const socklen_t &length() const         = 0;
+
+    struct sockaddr *sockaddr()
+    {
+        return const_cast<struct sockaddr *>(const_cast<const SockAddr *>(this)->sockaddr());
+    }
+    socklen_t &length()
+    {
+        return const_cast<socklen_t &>(const_cast<const SockAddr *>(this)->length());
+    }
+};
+
+/**
+ * Wrapper class around msghdr (structure describing messages sent by `sendmsg` and received by
+ * `recvmsg`)
+ */
+class MsgHdr {
+public:
+    virtual ~MsgHdr()                           = default;
+    virtual const struct msghdr *msghdr() const = 0;
+
+    struct msghdr *msghdr()
+    {
+        return const_cast<struct msghdr *>(const_cast<const MsgHdr *>(this)->msghdr());
+    }
+};
+
+/**
+ * Any object implementing the Socket interface is a FileDescriptor too.
+ * This way we give meaning to the action of "closing the socket".
+ */
+class Socket : public FileDescriptor {
+    virtual int setsockopt(int level, int optname, void *optval, socklen_t optlen) = 0;
+};
+
+/**
+ * Abstract base class for all types of sockets: Raw, UDP, TCP, UDS, ...
+ * This implementation class aggregates a FileDescriptor implementation so it has a file
+ * descriptor. Methods overridden from FileDescriptor interface delegate on the aggregated
+ * implementation.
+ * Derived classes provide the file descriptor obtained with a call to socket(), using different
+ * family, type and protocol parameters.
+ * This class aggregates a FileDescriptor instead of inheriting from one of its implementations to
+ * follow the principle of "Favor Aggregation over Inheritance".
+ * See https://wiki.c2.com/?UseCompositionAndInterfacesWithoutClassInheritance

```

Comment:
There is a mix here:
Since `Socket` _is a_ file descriptor as stated above, now `SocketAbstractImple` is **both** a `FileDescriptor` and has  a FileDescriptor (its `m_descriptor` member variable), which is completely confusing to me.

Reply:
Ok. As commented above, I'll try to use better names and also write better descriptions.